### PR TITLE
Fill Out Convert and Add Tests

### PR DESCRIPTION
**Pull request by Lixire**
Pertaining to issue #9328 
- Exposed new members
- Added new tests (Which run when using @weshaggard 's changes to the project.json and csproj file)

@ellismg I wasn't sure whether to include your changes for intellisense or not

@danmosemsft @joperezr

### CODE REVIEW

**Comment by ellismg** [comment]
> @ellismg I wasn't sure whether to include your changes for intellisense or not

I'd ask @weshaggard or @joperezr to review the commit (0ea6990) to make sure things are reasonable, but I see no reason to not include it.  Since I don't use this project in VS much I'm not sure if I fixed things the "right" way.

File: src/System.Runtime.Extensions/System.Runtime.Extensions.sln [hunk 1]
```diff
@@ -61,12 +71,16 @@ Global
 		{E73B023E-609E-4281-866A-3AECB1AB5D48}.Windows_Debug|Any CPU.Build.0 = Debug|Any CPU
 		{E73B023E-609E-4281-866A-3AECB1AB5D48}.Windows_Release|Any CPU.ActiveCfg = Debug|Any CPU
 		{E73B023E-609E-4281-866A-3AECB1AB5D48}.Windows_Release|Any CPU.Build.0 = Debug|Any CPU
-		{6C314C9B-3D28-4B05-9B4C-B57A00A9B3B9}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
-		{6C314C9B-3D28-4B05-9B4C-B57A00A9B3B9}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{6C314C9B-3D28-4B05-9B4C-B57A00A9B3B9}.Debug|Any CPU.ActiveCfg = Release|Any CPU
+		{6C314C9B-3D28-4B05-9B4C-B57A00A9B3B9}.Debug|Any CPU.Build.0 = Release|Any CPU
```
Comment (joperezr) [inline src/System.Runtime.Extensions/System.Runtime.Extensions.sln]: These two lines don't look right, I believe they should still be Debug

File: src/System.Runtime.Extensions/System.Runtime.Extensions.sln [hunk 2]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 2013
-VisualStudioVersion = 12.0.21005.1
+# Visual Studio 14
+VisualStudioVersion = 14.0.25420.1
```
Comment (joperezr) [inline src/System.Runtime.Extensions/System.Runtime.Extensions.sln]: I don't think it matters much, but I would undo these changes as they are irrelevant to your change.

File: src/System.Runtime.Extensions/tests/System.Runtime.Extensions.Tests.csproj [hunk 3]
```diff
@@ -16,10 +16,10 @@
     <NugetTargetMoniker Condition="'$(NugetTargetMoniker)'==''">.NETStandard,Version=v1.5</NugetTargetMoniker>
   </PropertyGroup>
```
Comment (joperezr) [inline src/System.Runtime.Extensions/tests/System.Runtime.Extensions.Tests.csproj]: If you want the tests to actually run against netstandard1.7 you will need to add an itemgroup here that looks like:
```xml
<ItemGroup Condition="'$(TargetGroup)'=='netstandard1.7'">
    <TestNugetTargetMoniker Include="$(NugetTargetMoniker)">
      <Folder>netcoreapp1.1</Folder>
    </TestNugetTargetMoniker>
  </ItemGroup>
```

Also you will need to add the following to your project.json in the supports section:
```
"coreFx.Test.netcoreapp1.1-ns17": {
        "netstandard1.7": [
            "win7-x86",
            "win7-x64",
            "win10-arm64",
            "osx.10.10-x64",
            "centos.7-x64",
            "debian.8-x64",
            "rhel.7-x64",
            "ubuntu.14.04-x64",
            "ubuntu.16.04-x64",
            "fedora.23-x64",
            "linux-x64",
            "opensuse.13.2-x64"
        ]
    }
```

File: src/System.Runtime.Extensions/tests/System/Convert.IsDBNull.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Xunit;
+
+namespace System.Tests
+{
+    public class IsDBNullTests
+    {
+#if netstandard17
```
Comment (joperezr) [inline src/System.Runtime.Extensions/tests/System/Convert.IsDBNull.cs]: You don't need these #ifdefs since this file is only includded on netstandard17

File: src/System.Runtime.Extensions/tests/System/Convert.ToBoolean.cs [hunk 5]
```diff
@@ -14,7 +14,6 @@ public void FromBoolean()
             Boolean[] testValues = { true, false };
             Verify(Convert.ToBoolean, testValues, testValues);
         }
-
         [Fact]
```
Comment (joperezr) [inline src/System.Runtime.Extensions/tests/System/Convert.ToBoolean.cs]: Nit: unwanted line removal.

File: src/System.Runtime.Extensions/tests/System/Convert.ToBoolean.cs [hunk 6]
```diff
@@ -124,5 +123,20 @@ public void FromUInt64()
             Boolean[] expectedValues = { false, true };
             Verify(Convert.ToBoolean, testValues, expectedValues);
         }
+#if netstandard17
```
Comment (joperezr) [inline src/System.Runtime.Extensions/tests/System/Convert.ToBoolean.cs]: Nit: you are also missing a new line before this #if's in some of your test files.

File: src/System.Runtime.Extensions/tests/System/Convert.ToString.cs [hunk 7]
```diff
@@ -651,6 +651,10 @@ public static void FromString()
                 result = Convert.ToString(testValues[i], NumberFormatInfo.CurrentInfo);
                 Assert.Equal(testValues[i].ToString(), result);
             }
+
+            string testValue = "Hello World!";
+            Assert.Equal(testValue, Convert.ToString(testValue));
+            Assert.Equal(testValue, Convert.ToString(testValue, NumberFormatInfo.CurrentInfo));
```
Comment (joperezr) [inline src/System.Runtime.Extensions/tests/System/Convert.ToString.cs]: Does this depend on new netstandard1.7 behavior?

**Comment by joperezr** [comment]
Overall, the changes looks good to me(baseline my previous comments), but I wonder if it would be better to add new files for all of these netstandard1.7 tests that we are adding in existing files instead of adding all of those #ifdefs which we usually try to avoid. @weshaggard thoughts?

File: src/System.Runtime.Extensions/tests/System/Convert.IsDBNull.cs [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    public class IsDBNullTests
+    {
+        [Fact]
+        public static void SimpleTest()
+        {
+            Assert.Equal(true, Convert.IsDBNull(Convert.DBNull));
+            Assert.Equal(false, Convert.IsDBNull(4));
+            Assert.Equal(false, Convert.IsDBNull(true));
+            Assert.Equal(false, Convert.IsDBNull('x'));
+            Assert.Equal(false, Convert.IsDBNull(1.1));
```
Comment (danmosemsft) [inline src/System.Runtime.Extensions/tests/System/Convert.IsDBNull.cs]: >Assert.Equal(false, Convert.IsDBNull(1.1)); [](start = 12, length = 43)

also
Assert.Equal(false, Convert.IsDBNull(null));

File: src/System.Runtime.Extensions/tests/System/Convert.ToBase64String.cs [hunk 9]
```diff
@@ -57,5 +57,20 @@ public static void InvalidLength()
...
+        {
+            byte[] barray = new byte[] { 1, 2, 3};
+            byte[] subset = new byte[] { 2, 3 };
+            string s1 = Convert.ToBase64String(barray, Base64FormattingOptions.InsertLineBreaks);
+            string s2 = Convert.ToBase64String(barray, Base64FormattingOptions.None);
+            string s3 = Convert.ToBase64String(barray, 1, 2, Base64FormattingOptions.None);
+            Assert.Equal(barray, Convert.FromBase64String(s1));
+            Assert.True(!s2.Contains("\n"));
+            Assert.Equal(subset, Convert.FromBase64String(s3));
+        }
```
Comment (danmosemsft) [inline src/System.Runtime.Extensions/tests/System/Convert.ToBase64String.cs]: >} [](start = 8, length = 1)

also
     Assert.Equal(barray, Convert.FromBase64String(s2));
?

File: src/System.Runtime.Extensions/tests/System/Convert.ToBase64CharArray.cs [hunk 10]
```diff
@@ -94,5 +94,16 @@ public static void InvalidInputLength()
...
+
+#if netstandard17
+        [Fact]
+        public static void SimpleTest()
+        {
+            byte[] barray = new byte[256];
+            char[] carray = new char[352];
+            int length = Convert.ToBase64CharArray(barray, 0, barray.Length, carray, 0, Base64FormattingOptions.InsertLineBreaks);
+            Assert.Equal(length, 352);
+        }
```
Comment (danmosemsft) [inline src/System.Runtime.Extensions/tests/System/Convert.ToBase64CharArray.cs]: >} [](start = 8, length = 1)

also
length = Convert.ToBase64CharArray(barray, 0, barray.Length, carray, 0, Base64FormattingOptions.None);
            Assert.Equal(length, 352);

**Comment by danmosemsft** [comment]
:shipit:

File: src/System.Runtime.Extensions/tests/System/Convert.IsDBNull.cs (hunk 8 above)
Comment (jasonwilliams200OK) [inline src/System.Runtime.Extensions/tests/System/Convert.IsDBNull.cs]: Perhaps `Assert.False` / `Assert.True` can be used instead.

File: src/System.Runtime.Extensions/tests/System.Runtime.Extensions.Tests.csproj [hunk 11]
```diff
@@ -31,6 +36,7 @@
     <Compile Include="System\BitConverter.cs" />
     <Compile Include="System\Convert.BoxedObjectCheck.cs" />
     <Compile Include="System\Convert.FromBase64.cs" />
+    <Compile Include="System\Convert.IsDBNull.cs" Condition="'$(TargetGroup)'=='netstandard1.7'" />
```
Comment (weshaggard) [inline src/System.Runtime.Extensions/tests/System.Runtime.Extensions.Tests.csproj]: Can you please create a separate TargetGroup conditioned on netstandard1.7?

File: src/System.Runtime.Extensions/tests/System/Convert.ToBase64CharArray.cs [hunk 12]
```diff
@@ -94,5 +94,18 @@ public static void InvalidInputLength()
             Assert.Throws<ArgumentOutOfRangeException>(() => Convert.ToBase64CharArray(inputBytes, 0, inputBytes.Length + 1, outputBuffer, 0));
             Assert.Throws<ArgumentOutOfRangeException>(() => Convert.ToBase64CharArray(inputBytes, 1, inputBytes.Length, outputBuffer, 0));
         }
+
+#if netstandard17
```
Comment (weshaggard) [inline src/System.Runtime.Extensions/tests/System/Convert.ToBase64CharArray.cs]: I would put this test in with the Dbunull tests and call it Convert.netstandard1.7

Comment (weshaggard) [inline src/System.Runtime.Extensions/tests/System/Convert.ToBase64CharArray.cs]: Similarly for all the @ifdef Convert tests through these files.

File: src/System.Runtime.Extensions/tests/System.Runtime.Extensions.Tests.csproj (hunk 11 above)
Comment (joperezr) [inline src/System.Runtime.Extensions/tests/System.Runtime.Extensions.Tests.csproj]: I suppose you meant ItemGroup?

Comment (weshaggard) [inline src/System.Runtime.Extensions/tests/System.Runtime.Extensions.Tests.csproj]: Yes a separate ItemGroup conditioned with TargetGroup netstandard1.7 :)

File: src/System.Runtime.Extensions/tests/System.Runtime.Extensions.Tests.csproj [hunk 13]
```diff
@@ -111,5 +111,11 @@
       <Name>VoidMainWithExitCodeApp</Name>
     </ProjectReference>
   </ItemGroup>
+  <!-- ToDo: Remove this target once the infrastructure for testing in netcoreapp1.1 is ready -->
+  <Target Name="RemoveUnwantedTestMonikers" Condition="'$(TargetGroup)'=='netstandard1.7'" BeforeTargets="CopyTestToTestDirectory">
```
Comment (weshaggard) [inline src/System.Runtime.Extensions/tests/System.Runtime.Extensions.Tests.csproj]: @joperezr I'm starting to get a little worried about the number of hacks we are introducing to get these tests running. Can we at a minimum put them in the root dir.targets so we don't have a bunch of these floating in all our test projects?

Comment (joperezr) [inline src/System.Runtime.Extensions/tests/System.Runtime.Extensions.Tests.csproj]: yeah that makes sense. Do you want that to happen as part of this PR or if not then I'll submit a new PR moving all of these to dir.targets once this is merged.

**Comment by danmosemsft** [comment]
@dotnet-bot test Innerloop OSX Debug Build and Test 
@dotnet-bot  test Innerloop Windows_NT Debug Build and Test

**Comment by Lixire** [comment]
@dotnet-bot test Innerloop Windows_NT Debug Build and Test
